testplan/runners/pools/base.py

```diff
@@ -471,27 +468,41 @@ def _workers_monitoring(self):
         else:
             loop_interval = 5  # seconds
 
-        while self.is_alive and self.active:
+        break_outer_loop = False
+        while self.active:
 
             hosts_status = {'active': [], 'inactive': [], 'initializing': []}
-            with self._pool_lock:
-                for worker in self._workers:
-                    status = self._query_worker_status(worker)
-                    hosts_status[status].append(worker)
-
-                if hosts_status != previous_status:
-                    self.logger.info('%s Hosts status update',
-                                     datetime.datetime.now())
-                    self.logger.info(pprint.pformat(hosts_status))
-                    previous_status = hosts_status
-
-                if not hosts_status['active'] \
-                        and not hosts_status['initializing'] \
-                        and hosts_status['inactive']:
-                    self.logger.critical(
-                        'All workers of {} are inactive.'.format(self))
-                    self.abort()
-                    break
+
+            for worker in self._workers:
+                status, reason = self._query_worker_status(worker)
+                if status == 'inactive':
+                    if self.active and self.status.tag not in (
+                            self.status.STOPPING, self.status.STOPPED):
+                        if self._handle_inactive(worker, reason):
```

Comment:
Handle inactive only if pool is not stopping

Reply:
There's still a window here where the status could become STOPPING before/during handling the inactive worker. Is that OK?